README
X-Kit - xorg.conf manipulator, parser and validator
=====================

X-Kit is a set of libraries to manipulate the xorg.conf. It was written
with the following aims in mind:

    * Simple API.
    * Easy to extend.
    * Dependent only on Python (no external libraries)

X-Kit was designed to be distribution agnostic so as to provide a
standard way to manipulate the content of the xorg.conf from different
applications.

Structure
---------

The Parser class in xorgparser.py is the core of the project since it
includes all the main methods for the manipulation and validation of
the xorg.conf. The XUtils subclass in xutils.py contains some not 
fundamental but still quite useful methods mainly focussed on dealing
with the several relationships between the different sections of the
xorg.conf.

The most important attribute of the Parser class is globaldict i.e. a
dictionary which contains a structured representation of the content
of the xorg.conf. The section types are its keys (see self.sections).
"SubSection" is treated as a special section type.

As regards normal sections, the value assigned to each key of 
globaldict is another dictionary which has the position of each section
as a key.

self.globaldict = {
                   'Device': {
                              0: [option1, option2, etc.],
                              1: [...], etc.
                             },
                   'Screen': {
                              0: [option1, option2, etc.],
                              1: [...], etc.},
                   etc.
                  }

For example self.globaldict['Device'][0] corresponds to a list of the
options available in the 1st "Device" section of the xorg.conf.

SubSections are treated differently from sections:

self.globaldict = {
                   'Device': {
                              0: [option1, option2, etc.],
                              1: [...], etc.
                             },
                   'Screen': {
                              0: [option1, option2, etc.],
                              1: [...], etc.
                             },
                             
                   'SubSection': {
                                  0: {
                                      'section': 'Screen',
                                      'identifier': 'Display',
                                      'position': 0,
                                      'options': [option1, option2, etc.]
                                     }, 
                                  1: {...}, etc.
                                 },
                  }

In this case they keys of self.globaldict['Device'] (i.e. 0, 1, etc.) 
represent the position of the "Display" SubSection inside the "Screen"
section. The following are the keys of self.globaldict['Device'][0], [1], etc.:

- section    = the section in which the SubSection is located (e.g. "Screen")
- position   = the position of the section (e.g. "Screen") in the xorg.conf
               e.g. position would be 0 if the SubSection was located in the
               1st "Screen" section of the xorg.conf (whose options are
               stored in "self.globaldict['Screen'][0]")
- identifier = the name of the SubSection (e.g. "Display")
- options    = a list with the options available in the SubSection

Another important concept to consider is that each line in each section is
considered as an option. No more than one instance of the same option should
be found in the same section (e.g. Driver "nv" and Driver "vesa" cannot 
cohexist in the same Device section). References (e.g. Screen "Default Screen"
in the Device section), however, are treated as a special kind of option and
more than one reference of the same kind can be found in the same section
(e.g. Screen "Screen 1" and Screen "Screen 2" in the same ServerLayout
section)

IMPORTANT NOTE ON CASE SENSITIVE SECTION NAMES: the parser will accept sections
in either lowercase and uppercase when reading a configuration file. The same
is not true when the methods are called and the name of the section is passed
as a parameter, as in such case the names of the sections must be exactly as in
"self.sections" i.e. the 1st letter of each word which is part of the name must 
be in uppercase:

e.g. invalid: "screen", "serverlayout", "Serverlayout"
     valid:   "Screen", "ServerLayout"

If you don't respect this rule, a KeyError will be raised.

For further information you can have a look at the documentation by typing:

import XKit.xutils
help(XKit.xutils)


Dependencies
------------

All that X-Kit needs is just a working Python installation. No external
modules are required.

How to install X-Kit
---------

If you use Ubuntu you should be able to find the source package on
Launchpad (see the "Development" section of this file) and in the
Ubuntu repository (starting from Ubuntu 8.10).

The easiest way to install it is by typing the following command:

    # ./setup.py install

    (as root).

Test suite
----------

Every new feature and bug fix should first get a test case, to
maintain nearly 100% code coverage. You can either run the 
following test files separately:

 tests/0-test.py = unit test for xorgparser.py
 tests/1-test.py = unit test for xutils.py

or simply run them all with the following file:

 tests/run 

The following file contains the path to the source xorg.conf which
will be used for testing ("inputFile") and the directory 
("outputDir") in which the results of the tests will be stored
("xkittest-0.txt" and "xkittest-1.txt" for 0-test.py and 1-test.py
respectively).

By default inputFile is set to the xorg.conf file in tests/ and
outputDir is set to tests/.

It is possible to change the default values of inputFile and 
outputDir by passing parameters to the "tests/run" file:

The only accepted (optional) parameters are:
  -o, --output=<dirname>    the directory where the results 
                            of the tests are saved.
  -i, --input=<filename>    the xorg.conf used for the tests.
  -h, --help                help page.

Launching "tests/run" without parameters will make it use the
default values for inputFile and outputDir.

Development
-----------

Development of trunk and the Ubuntu implementation takes place in
the following "x-kit" Launchpad project: 

  https://launchpad.net/x-kit
  
This is the place to report bugs, do translations, host your own
branches (if you wish to do so), and download upstream releases.

Alberto Milone (tseliot) <albertomilone@alice.it>


debian/changelog
x-kit (0.5.0) UNRELEASED; urgency=low

  * 

 -- Alberto Milone <alberto.milone@canonical.com>  Mon, 28 May 2012 17:42:55 +0200

x-kit (0.4.2.3build1) precise; urgency=low

  * Rebuild to drop python2.6 dependencies.

 -- Matthias Klose <doko@ubuntu.com>  Sat, 31 Dec 2011 02:15:46 +0000

x-kit (0.4.2.3) oneiric; urgency=low

  [ Alberto Milone ]
  * new upstream release: 0.4.2.3:
    + README:
      - Correct the instructions on how to get the documentation.
    + 1-example.py:
      - Use the right function to get the device position properly.

  [ Colin Watson ]
  * debian/control, debian/rules:
    - Convert to dh_python2 (LP: #788514).

 -- Alberto Milone <alberto.milone@canonical.com>  Tue, 28 Jun 2011 15:13:06 +0100

x-kit (0.4.2.2) lucid; urgency=low

  * new upstream release: 0.4.2.2:
    + xutils.py:
      - make isDriverInSection() return True only if all of the specified
        sections contain the driver. This prevents isDriverEnabled() from
        returning false positives.
    + 1-test.py:
      - Modify the test for testIsDriverEnabled() so as to test the correct
        behaviour.
      - Add two more tests for IsDriverEnabled().

 -- Alberto Milone <alberto.milone@canonical.com>  Tue, 30 Mar 2010 13:05:57 +0200

x-kit (0.4.2.1) lucid; urgency=low

  * new upstream release: 0.4.2.1:
    + xorgparser.py:
      - Add "InputClass" to the whitelist of xorg.conf sections so that
        the validator doesn't raise an error about it.

 -- Alberto Milone <alberto.milone@canonical.com>  Fri, 26 Mar 2010 13:46:06 +0100

x-kit (0.4.2) jaunty; urgency=low

  * new upstream release: 0.4.2:
    + xorgparser.py:
      - remove harmless but useless lines after an Exception in getValue()
      - make sure that if optiontype != None the option (differently from
        the value) is always put in quotation marks in addOption()
    + 0-test.py:
      - add a test for the quotation marks check in addOption()

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Sun, 04 Jan 2009 12:47:09 +0100

x-kit (0.4.1) UNRELEASED; urgency=low

  * new upstream release: 0.4.1:
    + xorgparser.py:
      - removeSection() and commentOutSection() now remove any reference to the
        sections they remove/comment out from self.identifiers
    + 0-test.py:
      - modify the existing tests to test the corrections to removeSection() and
        commentOutSection() in xorgparser.py

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Fri, 26 Dec 2008 17:10:03 +0100

x-kit (0.4) jaunty; urgency=low

  * new upstream release: 0.4:
    + xorgparser.py:
      - preserve commented lines inside sections too
      - add removeSection()
      - add commentOut{Option|SubOption|Section|SubSection}()
        o the new "Comments" key in self.globaldict contains commented options and
          subsections before these are merged into self.comments
        o when a whole section is commented out, its content and its subsections are
          immediately moved to self.comments
        o commented lines are not meant to be uncommented
        o commented lines are ignored by the validator
      - add getSubSections()
        o return the subsections contained in the section passed as an argument
      - separate __getOptionsToBlacklist() from removeOption() so that it can be shared
        with commentOutOption()
      - docstrings clean-up
    + 0-test.py:
      - add tests for the new methods in xorgparser.py

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Mon, 22 Dec 2008 16:56:51 +0100

x-kit (0.3.6) intrepid; urgency=low

  * xorgparser.py:
    - make sure that globaldict is preserved when
      writing its content to the destination file (use tempdict)
    - do not close file objects in writeFile() 
  * 0-test.py: add testWriteFile1(), testWriteFile2() and
    testWriteFile3()

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Wed, 10 Sep 2008 17:35:28 +0200

x-kit (0.3.5) intrepid; urgency=low

  * remove Replaces from the debian/control

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Sat, 16 Aug 2008 11:21:52 +0200

x-kit (0.3.4) intrepid; urgency=low

  [ Alberto Milone ]
  * add getBrokenReferences() to xorgparser.Parser()
  * move getPosition() and isSection() to xorgparser since they are now required
    by getBrokenReferences(). Move the relevant tests to 0-test.py
  * remove all the arguments from xutils.XUtils.checkNFixSection()
  * make checkNFixSection() rely on getBrokenReferences()
  * Parser.__process() will now check the following additional cases:
    - if duplicate sections are found
    - if broken references are found
    - if there are sections which don't have an identifier but should
      have one (except for completely empty sections, which are ignored
      and don't appear in globaldict)
  * clean up Parser.getIds()
  * remove sys.exit from Parser.getReferences()
  * disable checkDuplicateOptions by default
  * move getIdentifier and isSection to xorgparser
  * move all the validator methods to __complianceRules()
  * add getBrokenReferences() to xorgparser
  * add validateOptions to the default checks
  * improve getDefaultServerLayout() in xorgparser so that if one default ServerLayout
    is set, its existence will be checked
  * make sure that each ServerLayout has at least 1 reference to a Screen section
  * print something useful for debugging when raising ParseException
  * update the README with a note on case-sensitive section names

  [ Martin Pitt ]
  * debian/rules: Remove some cruft.
  * debian/control: Rename package x-kit to python-xkit for conforming to
    Python package naming policy. Also fix section to be 'python'.
  * debian/rules: Run the test suite during build and clean up the spewage.
  * Add debian/python-xkit.examples: Ship examples/ directory.
  * debian/control: Add Vcs-Bzr:.

 -- Alberto Milone (tseliot) <albertomilone@alice.it>  Thu, 14 Aug 2008 18:36:12 +0200

x-kit (0.3.3) intrepid; urgency=low


xkit/xutils.py
#       xutils.py -- Enhanced class of X-Kit's parser
#       
#       Copyright 2008 Alberto Milone <albertomilone@alice.it>
#       
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#       
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#       
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.

from .xorgparser import *
import sys

class XUtils(Parser):
    '''Subclass with higher-level methods
    
    See xorgparser.Parser for the low-level methods'''
    def __init__(self, source=None):
        super(XUtils, self).__init__(source)
    
    def fix_broken_references(self):
        '''Fix broken references to non-existent sections'''
        broken_references = self.get_broken_references()
        for section in broken_references:
            for reference in broken_references[section]:
                self.make_section(section, identifier=reference)
     
    def get_driver(self, section, position):
        '''Get the driver in use in a section

        If no driver is found it will return False.

        For further information see get_value()'''
        option = 'Driver'
        return self.get_value(section, option, position)
        
    def set_driver(self, section, driver, position):
        '''Set the driver in use in a section'''
        option = 'Driver'
        self.add_option(section, option, driver, position=position)

    def section_has_driver(self, driver, sections_list=None):
        '''Look for a driver in the Device sections

        Return True if the driver is found in each of the specified
        sections, otherwise return False.

        if sections_list == None check all the Device sections'''
        if sections_list == None:
            sections_list = list(self.globaldict['Device'].keys())

        for section in sections_list:
            try:
                if self.get_driver('Device', section) != driver:
                    return False
            except OptionException:
                #no references to the Device section
                return False
        return True

    def get_devices_in_serverlayout(self, position):
        '''Return a list of references to the Device sections in ServerLayout
        
        This method looks for references to Device sections in the Screen
        sections referred to in the ServerLayout[position] section.'''
        devices_to_check = []
        references = self.get_references('ServerLayout', position, ['Screen'])
        if len(references['Screen']) > 0:
            # Check all the device sections related to these Screen sections
            #
            # references will look like {'Screen': ['Screen1', '0']}
            for reference in references['Screen']:
                try:
                    screen_position = self.get_position('Screen', reference)
                except IdentifierException:
                    continue
                # Get references to the Device sections in the Screen sections
                try:
                    device_references = self.get_references('Screen',
                                             screen_position, ['Device'])
                    for device in device_references['Device']:
                        device_position = self.get_position('Device', device)
                        devices_to_check.append(device_position)
                except OptionException:
                    #no references to the Device section
                    pass
        return devices_to_check
    
    def get_devices_in_use(self):
        '''Return the Device sections in use

        If no Device sections are referenced in ServerLayout then all of
        the available Device sections are returned.

        This method supports old Xinerama setups and therefore looks for
        references to Device sections in the ServerLayout section(s) and
        checks only the default ServerLayout section provided than one is
        set in the ServerFlags section.'''
        devices_to_check = []
        driver_enabled = False
        
        serverlayout = self.globaldict['ServerLayout']
        serverflags = self.globaldict['ServerFlags']
        serverlayout_length = len(serverlayout)
        serverflags_length = len(serverflags)
        
        if serverlayout_length > 0:
            if serverlayout_length > 1:#More than 1 ServerLayout?
                if serverflags_length > 0:#has ServerFlags
                    # If the ServerFlags section exists there is a chance that
                    # a default ServerLayout is set.
                    #
                    # If no ServerLayout is set, this might be intentional
                    # since the user might start X with the -layout command
                    # line option.

                    # See if it has a default ServerLayout
                    default = self.get_default_serverlayout()

                    if len(default) == 1:
                        devices_to_check = \
                        self.get_devices_in_serverlayout(default[0])
                    else:
                        for layout in serverlayout:
                            devices_to_check += \
                            self.get_devices_in_serverlayout(layout)
                else:
                    for layout in serverlayout:
                        devices_to_check += \
                        self.get_devices_in_serverlayout(layout)
            else:
                devices_to_check = self.get_devices_in_serverlayout(0)

        if len(devices_to_check) == 0:
            # Check all the Device sections
            devices_to_check = list(self.globaldict['Device'].keys())

        return devices_to_check
    
    def is_driver_enabled(self, driver):
        '''See if a driver is enabled in the Device sections

        When possible, this method checks only the Device sections in use,
        otherwise it checks any available Device section.

        This method supports old Xinerama setups and therefore looks for
        references to Device sections in the ServerLayout section(s) and
        checks only the default ServerLayout section provided than one is
        set in the ServerFlags section.'''
        devices_to_check = self.get_devices_in_use()
        driver_enabled = self.section_has_driver(driver,
                                               sections_list=devices_to_check)

        return driver_enabled

    def get_screen_device_relationships(self):
        '''See which Screen sections are related to which Device sections'''
        relationships = {}
        it = 0
        for screen in self.globaldict['Screen']:
            references = self.get_references('Screen', it, reflist=['Device'])
            device = references['Device'][0]
            device = self.get_position('Device', device)
            relationships.setdefault(device)
            relationships[device] = {}
            relationships[device]['Screen'] = it
            it += 1

        return relationships



examples/0-example.py
#       0-example.py -- Examples which rely on xorgparser
#       
#       Copyright 2008 Alberto Milone <albertomilone@alice.it>
#       
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#       
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#       
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.

from xkit import xorgparser
from xkit.xorgparser import *
import sys
import os

def main():
    '''
    Replace the first line of this example with a source and a destination file
    '''
    destination = os.path.join(os.path.expanduser('~'), 'xorgnew.txt')
    #source = '/etc/X11/xorg.conf'
    source = None
    a = Parser(source)
    
    '''
    Remove the Defaultdepth from all the Screen sections
    '''
    a.remove_option('Screen', 'DefaultDepth')
    
    '''
    Remove the Defaultdepth from the 1st Screen section
    '''
    a.remove_option('Screen', 'DefaultDepth', position=0)
    
    '''
    Set the Defaultdepth to 24 bit in all sections
    '''
    a.add_option('Screen', 'Defaultdepth', '24')
    '''
    Set the Defaultdepth to 24 bit in the 1st Screen section
    '''
    a.add_option('Screen', 'Defaultdepth', '24', position=0, prefix='')
    
    '''
    Get the value assigned to the AddARGBGLXVisuals option in the 1st Screen section
    '''
    try:
        print('AddARGBGLXVisuals', a.get_value('Screen', 'AddARGBGLXVisuals', position=1, identifier='Display'))
    except OptionException as e:
        print('Error:', e)
    
    '''
    Add an Option (with the Option prefix) to the 1st Screen section
    '''
    a.add_option('Screen', 'AddARGBGLXVisuals', 'True', option_type='Option', position=0)
    
    '''
    Print the global dict i.e. the dict which contains all sections
    '''
    print('\nGlobal dict is the dict which contains all sections\n', a.globaldict)
    #print '\nGlobal iters is\n', a.globaliters
    
    '''
    Make a new Device and a Screen section with an identifier
    '''
    a.make_section('Device', identifier='Name of this new Device Section')
    a.make_section('Screen', identifier='Default Screen')
    
    '''
    Add a reference to the Screen section identified as "New Default Screen" to
    the 1st ServerLayout section
    '''
    print('\nNew Default Screen in progress')
    a.add_reference('ServerLayout', 'Screen', 'New Default Screen', position=0)
    a.add_reference('ServerLayout', 'InputDevice', 'New Device', position=0)
    a.remove_reference('ServerLayout', 'InputDevice', 'New Device', position=0)
    '''
    Add a reference to the Screen section identified as "New Default Screen" to
    all the ServerLayout sections
    '''
    #a.add_reference('ServerLayout', 'Screen', 'New Default Screen', position=0)
    
    '''
    Remove a reference to the Screen section identified as "New Default Screen"
    from all the ServerLayout sections
    '''
    #a.remove_reference('ServerLayout', 'Screen', 'New Default Screen')#, position=0)
    
    '''
    Create a new "Display" SubSection inside all the Screen sections
    '''
    a.make_subsection('Screen', 'Display')#, position=0)
    '''
    Remove a "Display" SubSection inside all the Screen sections
    '''
    #a.remove_subsection('Screen', 'Display')#, position=0)
    
    '''
    Add an option to the Display subsection of the 1st Screen section
    '''
    a.add_suboption('Screen', 'Display', 'Depth', value='24', position=0)
    a.add_suboption('Screen', 'Display', 'Virtual', value='1600 1200', position=0)
    a.add_suboption('Screen', 'Display', 'Name', value='Whatever', option_type='Option', position=None)
    
    '''
    Remove options from the Display subsection of all or of the 1st Screen section
    '''
    a.remove_suboption('Screen', 'Display', 'Depth')
    a.remove_suboption('Screen', 'Martin', 'Virtual', position=0)
    
    
    '''
    Get the identifier of the 1st Device section
    '''
    print(a.get_value('Device', 'Identifier', 0))
    #print a.get_value('SubSection', 'Name', position=0, identifier='Display', sect='Screen')
    
    
    a.make_section('Screen', 'New Screen')
    a.make_section('Screen', 'New Screen')#this new section won't be created
    
    '''
    Create a new device section
    add a new option to it
    and make a reference to it in the Screen section
    '''
    dev = a.make_section('Device', 'My Device')
    a.add_option('Device', 'BusID', 'PCI:1:0:0', position=dev)
    a.add_reference('Screen', 'Device', 'My Device', position=0)
    
    a.add_reference('Device', 'Screen', 4, position=0)
    print(a.get_references('Screen', 0, reflist=['Device']))
    
    print('Virtual', a.get_value('SubSection', 'Virtual', position=0, identifier='Display', sect='Screen'))
    print('Modes', a.get_value('SubSection', 'Modes', position=0, identifier='Display', sect='Screen'))
    
    '''
    Create a new Device section and print the list of identifiers so as to see
    that the new identifier and position are included in identifiers
    '''
    a.make_section('Device', identifier='New Graphics Card')
    
    print('\nCreate Broken Screen section')
    pos = a.make_section('Screen', identifier='Broken Screen Section')
    print('\nAdding References')
    a.add_reference('Screen', 'Monitor', 'Broken Monitor Section', position=pos)
    a.add_reference('Screen', 'Device', 'Broken Device Section', position=pos)
    
    try:
        print('Horizsync value is', a.get_value('Monitor', 'Horizsync', 0))
    except OptionException as e:
        print('Error:', e)
    
    '''
    Write the changes to the destination file
    '''
    a.write(destination)

if __name__ == '__main__': main()


examples/1-example.py
#       1-example.py -- Examples which rely on accessories 
#                       (and therefore also on xutils)
#       
#       Copyright 2008 Alberto Milone <albertomilone@alice.it>
#       
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#       
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#       
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#       MA 02110-1301, USA.

from .accessories import *
import os

def main():
    '''
    Replace the first line of this example with a source and a destination file
    '''
    destination = os.path.join(os.path.expanduser('~'), 'xorgnew.txt')
    source = None
    a = Accessories(source)
    
    '''
    Remove the Defaultdepth from all the Screen sections
    '''
    a.remove_option('Screen', 'DefaultDepth')
    
    '''
    Remove the Defaultdepth from the 1st Screen section
    '''
    a.remove_option('Screen', 'DefaultDepth', position=0)
    
    '''
    Set the Defaultdepth to 24 bit in all sections
    '''
    a.add_option('Screen', 'Defaultdepth', '24', prefix='')
    '''
    Set the Defaultdepth to 24 bit in the 1st Screen section
    '''
    a.add_option('Screen', 'Defaultdepth', '24', position=0, prefix='')
    
    '''
    Get the value assigned to the AddARGBGLXVisuals option in the 1st Screen section
    '''
    try:
        print('AddARGBGLXVisuals', a.get_value('Screen', 'AddARGBGLXVisuals', position=1, identifier='Display'))
    except OptionNotAvailableException as e:
        print('Error:', e)
    '''
    Add an Option (with the Option prefix) to the 1st Screen section
    '''
    a.add_option('Screen', 'AddARGBGLXVisuals', 'True', option_type='Option', position=0)
    
    '''
    Print the contents of all the Device sections
    '''
    a.printSection('Device')
    
    '''
    Print the global dict i.e. the dict which contains all sections
    '''
    print('\nGlobal dict is the dict which contains all sections\n', a.globaldict)
    #print '\nGlobal iters is\n', a.globaliters
    
    '''
    Make a new Device and a Screen section with an identifier
    '''
    a.make_section('Device', identifier='Name of this new Device Section')
    a.make_section('Screen', identifier='Default Screen')
    
    '''
    Add a reference to the Screen section identified as "New Default Screen" to
    the 1st ServerLayout section
    '''
    a.add_reference('ServerLayout', 'Screen', 'New Default Screen')
    '''
    Add a reference to the Screen section identified as "New Default Screen" to
    all the ServerLayout sections
    '''
    a.add_reference('ServerLayout', 'Screen', 'New Default Screen', position=0)
    
    '''
    Remove a reference to the Screen section identified as "New Default Screen"
    from all the ServerLayout sections
    '''
    a.remove_reference('ServerLayout', 'Screen', 'New Default Screen')#, position=0)
    
    '''
    Create a new "Display" SubSection inside all the Screen sections
    '''
    a.make_subsection('Screen', 'Display')#, position=0)
    '''
    Remove a "Display" SubSection inside all the Screen sections
    '''
    #a.remove_subsection('Screen', 'Display')#, position=0)
    
    '''
    Add an option to the Display subsection of the 1st Screen section
    '''
    a.add_suboption('Screen', 'Display', 'Depth', value='24', position=0, prefix='')
    a.add_suboption('Screen', 'Display', 'Virtual', value='1600 1200', position=0)
    a.add_suboption('Screen', 'Display', 'Name', value='Whatever', option_type='Option', position=None)
    
    '''
    Remove options from the Display subsection of all or of the 1st Screen section
    '''
    a.remove_suboption('Screen', 'Display', 'Depth')
    a.remove_suboption('Screen', 'Martin', 'Virtual', position=0)
    
    
    '''
    Get the identifier of the 1st Device section
    '''
    print(a.get_value('Device', 'Identifier', 0))
    #print a.get_value('SubSection', 'Name', position=0, identifier='Display', sect='Screen')
    
    
    '''
    Set the driver  of the 1st Device section
    '''
    a.set_driver('Device', 'fbdev', 0)
    '''
    Get the driver  of the 1st Device section
    '''
    print(a.get_driver('Device', 0))
    
    
    a.make_section('Screen', 'New Screen')
    a.make_section('Screen', 'New Screen')#this new section won't be created
    a.setDefaultDepth(24, 0)
    print(a.getDefaultDepth(0))
    
    '''
    Create a new device section
    add a new option to it
    and make a reference to it in the Screen section
    '''
    dev = a.make_section('Device', 'My Device')
    a.add_option('Device', 'BusID', 'PCI:1:0:0', position=dev)
    a.add_reference('Screen', 'Device', 'My Device', position=0)
    
    a.add_reference('Device', 'Screen', 4, position=0)
    print(a.get_references('Screen', 0, reflist=['Device']))
    a.enableComposite()
    a.addArgbGlxVisuals(0)
    
    print('Virtual', a.get_value('SubSection', 'Virtual', position=0, identifier='Display', sect='Screen'))
    print('Modes', a.get_value('SubSection', 'Modes', position=0, identifier='Display', sect='Screen'))
    
    '''
    Get the identifier of the first Device section
    '''
    print('ID of the 1st Device Section =', a.get_identifier('Device', 0))
    
    '''
    Get the position of the Device section identified as 'Configured Video Device'
    '''
    try:
        print('Position of "Configured Video Device" =', a.get_position('Device', 'Configured Video Device'))
    except IdentifierException as e:
        print(e)
    '''
    See if a section exists
    '''
    print('Section Device "Configured Video Device" exists =', a.is_section('Device', 'Configured Video Device'))
    print('Section Device "Whatever" exists =', a.is_section('Device', 'Whatever'))
    
    '''
    Create a new Device section and print the list of identifiers so as to see
    that the new identifier and position are included in identifiers
    '''
    a.make_section('Device', identifier='New Graphics Card')
    a.make_section('Screen', identifier='New Screeeeeeeeeen')
    print('\nIdentifiers after creating a new device section', a.identifiers)
    
    print('\nCreate Broken Screen section')
    pos = a.make_section('Screen', identifier='Broken Screen Section')
    print('\nAdding References')
    a.add_reference('Screen', 'Monitor', 'Broken Monitor Section', position=pos)
    a.add_reference('Screen', 'Device', 'Broken Device Section', position=pos)
    
    '''
    Try to fix section with broken references
    '''
    a.fix_broken_references()
    '''
    Write the changes to the destination file
    '''
    a.write(destination)

if __name__ == '__main__': main()
